### PR TITLE
Take account of masks when summing bins

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
Solves #1677. See updated test for example of new functionality.

### CODE REVIEW

File: dataset/variable_reduction.cpp [hunk 1]
```diff
@@ -16,8 +16,7 @@
 namespace scipp::dataset {
 namespace {
 Variable applyMask(const VariableConstView &var, const Variable &masks) {
-  return scipp::variable::transform(var, masks,
-                                    scipp::core::element::convertMaskedToZero);
+  return scipp::variable::masked_to_zero(var, masks);
 }
```
Comment (SimonHeybrock) [inline dataset/variable_reduction.cpp:20]: Why the trivial wrapper function?

Comment (Matthew-Andrew) [reply dataset/variable_reduction.cpp:20]: Must have misunderstood your previous comment. Were you suggesting moving the whole apply mask functionality to a common location instead?

Comment (SimonHeybrock) [reply dataset/variable_reduction.cpp:20]: I think you understood correctly, I am just saying that you don't need to define `applyMask`, but just `masked_to_zero` directly?

Comment (Matthew-Andrew) [reply dataset/variable_reduction.cpp:20]: Good point will remove.

File: dataset/variable_reduction.cpp [hunk 2]
```diff
@@ -65,8 +60,10 @@ Variable nanmean(const VariableConstView &var, const Dim dim,
                  const MasksConstView &masks) {
   using variable::isfinite;
   if (const auto mask_union = irreducible_mask(masks, dim)) {
-    const auto count = sum(applyMask(isfinite(var), mask_union), dim);
-    return nanmean_impl(applyMask(var, mask_union), dim, count);
+    const auto count =
+        sum(scipp::variable::masked_to_zero(isfinite(var), mask_union), dim);
```
Comment (SimonHeybrock) [inline dataset/variable_reduction.cpp:64]: You don't need the `scipp::variable` here or in any of the other cases. ADL will take care of the test.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**